src/runtime/hexagon_remote/halide_host.cpp

```diff
@@ -1,23 +1,264 @@
-#include <rpcmem.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <linux/types.h>
+#include <android/log.h>
+
+#include "bin/src/halide_hexagon_remote.h"
+
+#include "stdio.h"
+
+namespace {
+
+typedef halide_hexagon_remote_handle_t handle_t;
+typedef halide_hexagon_remote_buffer buffer;
+
+// Allocations that are intended to be shared with Hexagon can be
+// shared without copying if they are contiguous in physical
+// memory. Android's ION allocator gives us a mechanism with which we
+// can allocate contiguous physical memory.
+enum ion_heap_id {
+    system_heap_id = 25,
+};
+
+enum ion_flags {
+    ion_flag_cached = 1,
+};
+
+typedef int ion_user_handle_t;
+
+struct ion_allocation_data {
+    size_t len;
+    size_t align;
+    unsigned int heap_id_mask;
+    unsigned int flags;
+    ion_user_handle_t handle;
+};
+
+struct ion_fd_data {
+    ion_user_handle_t handle;
+    int fd;
+};
+
+struct ion_handle_data {
+    ion_user_handle_t handle;
+};
+
+#define ION_IOC_ALLOC _IOWR('I', 0, ion_allocation_data)
+#define ION_IOC_FREE _IOWR('I', 1, ion_handle_data)
+#define ION_IOC_MAP _IOWR('I', 2, ion_fd_data)
+
+ion_user_handle_t ion_alloc(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags) {
+    ion_allocation_data alloc = {
+        len,
+        align,
+        heap_id_mask,
+        flags,
+        0
+    };
+    if (ioctl(ion_fd, ION_IOC_ALLOC, &alloc) < 0) {
+        return -1;
+    }
+    return alloc.handle;
+}
+
+int ion_map(int ion_fd, ion_user_handle_t handle) {
+    ion_fd_data data = {
+        handle,
+        0
+    };
+    if (ioctl(ion_fd, ION_IOC_MAP, &data) < 0) {
+        return -1;
+    }
+    return data.fd;
+}
+
+int ion_free(int ion_fd, ion_user_handle_t ion_handle) {
+    if(ioctl(ion_fd, ION_IOC_FREE, &ion_handle) < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+// We need to be able to keep track of the size and some other
+// information about ION allocations, so define a simple linked list
+// of allocations we can traverse later.
+struct allocation_record {
+    allocation_record *next;
+    int ion_fd;
+    ion_user_handle_t handle;
+    int buf_fd;
+    void *buf;
+    size_t size;
+};
+
+// Make a dummy allocation so we don't need a special case for the
+// head list node.
+allocation_record allocations = { NULL, };
```

Comment:
It's weird that this is a dummy allocation recorder instead of just a pointer. I think if the list removal code tracked the address of the prev pointer (allocation_record **), instead of the address of the prev record (allocation_record *), things could be slightly simpler.